velox/expression/ComplexProxyTypes.h

```diff
@@ -0,0 +1,220 @@
+  FOLLY_ALWAYS_INLINE element_t& add_item() {
+    commitMostRecentChildItem();
+    auto index = valuesOffset_ + length_;
+    reserve(index + 1);
+    length_++;
+
+    if constexpr (!provide_std_interface) {
+      childWriter_->setOffset(index);
+      needCommit_ = true;
+      return childWriter_->current();
+    } else {
+      childWriter_->vector().setNull(index, false);
+      return childWriter_->data_[index];
+    }
+  }
+
+  // Add a new null item to the array.
+  FOLLY_ALWAYS_INLINE void add_null() {
+    commitMostRecentChildItem();
+    reserve(valuesOffset_ + length_ + 1);
+    childWriter_->vector().setNull(valuesOffset_ + length_, true);
+    length_++;
+    // Note: no need to commit the null item.
+  }
+
+  // Should be called by the user (VectorWriter) when writing is done to commit
+  // last item if needed.
+  void finalize() {
+    commitMostRecentChildItem();
+    // Downsize to the actual size used in the underlying vector.
+    // Some vector-writer's logic depend on the previous size to append data.
+    childWriter_->vector().resize(valuesOffset_ + length_);
+  }
+
+  vector_size_t size() {
+    return length_;
+  }
+
+  // Functions below provides an std::like interface, and are enabled only when
+  // the array element is primitive that is not string or bool.
+
+  // 'size' is with respect to the current size of the array being written.
+  FOLLY_ALWAYS_INLINE typename std::enable_if<provide_std_interface>::type
+  resize(vector_size_t size) {
+    commitMostRecentChildItem();
+    reserve(size);
+    length_ = size;
+  }
+
+  typename std::enable_if<provide_std_interface>::type FOLLY_ALWAYS_INLINE
+  push_back(element_t value) {
+    auto& item = add_item();
+    item = value;
+  }
+
+  typename std::enable_if<provide_std_interface>::type FOLLY_ALWAYS_INLINE
+  push_back(std::nullopt_t) {
+    add_null();
+  }
+
+  typename std::enable_if<provide_std_interface>::type FOLLY_ALWAYS_INLINE
+  push_back(const std::optional<element_t>& value) {
+    if (value) {
+      push_back(*value);
+    } else {
+      add_null();
+    }
+  }
+
+  typename std::enable_if<provide_std_interface, PrimitiveWriterProxy<V>>::type
+  operator[](vector_size_t index_) {
+    return PrimitiveWriterProxy<V>{
+        &childWriter_->vector(), valuesOffset_ + index_};
+  }
+
+ private:
+  ArrayProxy<V>() {}
+
+  // Note: size is with respect to the current size of this array being written.
+  FOLLY_ALWAYS_INLINE void reserve(vector_size_t size) {
+    if (UNLIKELY(size > capacity_)) {
+      while (capacity_ < size) {
+        capacity_ = 1.5 * capacity_ + 1;
+      }
+      childWriter_->ensureSize(valuesOffset_ + capacity_);
+    }
+  }
+
+  FOLLY_ALWAYS_INLINE void commitMostRecentChildItem() {
+    if constexpr (!provide_std_interface) {
+      if (needCommit_) {
+        childWriter_->commit(true);
+        needCommit_ = false;
+      }
+    }
+  }
+
+  // Prepare the proxy for a new element.
+  // TODO: childWriter does not change.
```

Comment:
What is this TODO about?

Reply:
avoid passing childWriter every time we call init()

Reply:
I just fix it.